libsplinter/src/admin/store/diesel/models.rs

```diff
@@ -0,0 +1,569 @@
+                        circuit_id: proposed_circuit.circuit_id.clone(),
+                        node_id: node.node_id.clone(),
+                        endpoint: endpoint.clone(),
+                    })
+                    .collect::<Vec<ProposedNodeEndpointsModel>>(),
+            );
+        }
+        endpoint_models
+    }
+}
+
+/// Database model representation of a `ProposedService`
+#[derive(Debug, PartialEq, Associations, Identifiable, Insertable, Queryable, QueryableByName)]
+#[table_name = "proposed_services"]
+#[belongs_to(ProposedCircuitModel, foreign_key = "circuit_id")]
+#[primary_key(circuit_id, service_id)]
+pub struct ProposedServiceModel {
+    pub circuit_id: String,
+    pub service_id: String,
+    pub service_type: String,
+}
+
+impl From<&ProposedCircuit> for Vec<ProposedServiceModel> {
+    fn from(proposed_circuit: &ProposedCircuit) -> Self {
+        proposed_circuit
+            .roster
+            .iter()
+            .map(|service| ProposedServiceModel {
+                circuit_id: proposed_circuit.circuit_id.clone(),
+                service_id: service.service_id.clone(),
+                service_type: service.service_type.clone(),
+            })
+            .collect()
+    }
+}
+
+/// Database model representation of the `allowed_nodes` associated with a `ProposedService`
+#[derive(Debug, PartialEq, Associations, Identifiable, Insertable, Queryable, QueryableByName)]
+#[table_name = "proposed_services_allowed_nodes"]
+#[belongs_to(ProposedCircuitModel, foreign_key = "circuit_id")]
+#[belongs_to(ProposedServiceModel, foreign_key = "service_id")]
+#[primary_key(circuit_id, service_id, allowed_node)]
+pub struct ProposedServiceAllowedNodesModel {
+    pub circuit_id: String,
+    pub service_id: String,
+    pub allowed_node: String,
+}
+
+impl From<&ProposedCircuit> for Vec<ProposedServiceAllowedNodesModel> {
+    fn from(proposed_circuit: &ProposedCircuit) -> Self {
+        let mut allowed_nodes = Vec::new();
+        for service in &proposed_circuit.roster {
+            allowed_nodes.extend(
+                service
+                    .allowed_nodes
+                    .iter()
+                    .map(|node| ProposedServiceAllowedNodesModel {
+                        circuit_id: proposed_circuit.circuit_id.clone(),
+                        service_id: service.service_id.clone(),
+                        allowed_node: node.clone(),
+                    })
+                    .collect::<Vec<ProposedServiceAllowedNodesModel>>(),
+            );
+        }
+        allowed_nodes
+    }
+}
+
+/// Database model representation of the arguments associated with a `ProposedService`
+#[derive(Debug, PartialEq, Associations, Identifiable, Insertable, Queryable, QueryableByName)]
+#[table_name = "proposed_services_arguments"]
+#[belongs_to(ProposedCircuitModel, foreign_key = "circuit_id")]
+#[belongs_to(ProposedServiceModel, foreign_key = "service_id")]
+#[primary_key(circuit_id, service_id)]
```

Comment:
The primary key should include the `key` field since there can be multiple arguments for each (circuit_id, service_id). Same for non-proposal model.